design-documents/testing/functional/versioning-and-backward-compatibility-policy.md

```diff
@@ -28,82 +31,75 @@ An approach of defining what each release should include was taken from [Semanti
+Minor version **Y** MUST be incremented if a new, backwards compatible test or test entity is introduced.
+It MUST be incremented if any test or test entity is marked as `deprecated`.
+It MAY include patch level changes. Patch version MUST be reset to 0 when the minor version is incremented.
+
+### X release
+
+Major version **X** MUST be incremented if any backwards incompatible changes are introduced to a test or test entity.
+It MAY include minor and patch level changes. Patch and minor version MUST be reset to 0 when the major version is incremented.
 
 ## Implementation tasks
-1. Add Semantic Version analyzer to be able automatically define release type of MFTF tests package.
+
+1. Add Semantic Version analyzer to be able automatically define the release type of the MFTF tests package.
 2. Update publication infrastructure to exclude tests from `magento2-module` package type.
 3. Introduce publication functionality for publishing `magento2-test-module` package type.
-4. Create metapackage with test packages only for each Magento edition.
+4. Create a metapackage with test packages specifically for each Magento edition.
 
-## Entity's attributes
-Changing any of this attribute will cause Backward Incompatible change.
-
-**Test entity:**
-  
-  |xPath|attribute|
-  |---|---|
-  |`/tests/test`|name|
-  |`/tests/test/<ACTION> ⃰`|stepKey|
-  |`/tests/test/before/<ACTION> ⃰`|stepKey|
-  |`/tests/test/after/<ACTION> ⃰`|stepKey| 
-
-**Action Group entity:**
-
-  |xPath|attribute|
-  |---|---|
-  |`/actionGroups/actionGroup`|name|
-  |`/actionGroups/actionGroup/arguments/argument`|name|
-  |`/actionGroups/actionGroup/arguments/argument`|type|
-  |`/actionGroups/actionGroup/<ACTION> ⃰`|stepKey|
-
-**Data entity:**
-  
-  |xPath|attribute|
-  |---|---|
-  |`/entities/entitie`|name|
-  |`/entities/entitie/data`|key|
-  |`/entities/entitie/array`|key|
-  |`/entities/entitie/var`|key|
-  |`/entities/entitie/requiredEntity`|type|
-
-**Metadata entity:**
-  
-  |xPath|attribute|
-  |---|---|
-  |`/operations/operation`|name|
-  |`/operations/operation/field`|key|
-  |`/operations/operation(/object)+`|key|
-  |`/operations/operation(/object)+/field`|key|
-  |`/operations/operation(/object)+/array`|key|
-  |`/operations/operation/array`|key|
-  |`/operations/operation/array/object`|key|
+## Version increase matrix
   
-**Page entity:**
-
-  |xPath|attribute|
-  |---|---|
-  |`/pages/page`|name|
-  |`/pages/page/section`|name|
-  
-**Section entity:**
-
-  |xPath|attribute|
-  |---|---|
-  |`/sections/section`|name|
-  |`/sections/section/element`|name|
+|Entity Type|Change|Version Increase|
+|---|---|---|
+|ActionGroup|`<actionGroup>` added|MINOR
+| |`<actionGroup>` removed|MAJOR
+| |`<actionGroup>` `<action>` added|MINOR
+| |`<actionGroup>` `<action>` removed|MAJOR
+| |`<actionGroup>` `<action>` type changed|PATCH
+| |`<actionGroup>` `<action>` attribute changed|PATCH
```

Comment:
Which attribute? Might be minor if `ref` change or do we cover it separately.

Reply:
@tomreece raised the same concern, `ref` means an actionGroup may change which can introduce new test actions. I think it can be changed to the following:
* `ref` => `MINOR` (since ext devs only care about merging before or after this set of actions, but the actions **do** change)
* `all others except for stepKey` => `PATCH` (since it will not affect merging).

Reply:
Can it affect flow of the test and make test that depends on the flow fail (I don't know what are the other attributes)?

Reply:
In the situation that:
```
ActionGroup1 contains AmOnPage, Click, FillField
ActionGroup2 contains AmOnPage, Click, FillField, Click(submit button)
```

Then yes, changing actionGroups can alter test flow. If my extension merges after this `actionGroup` call and relies on one more `fillField` before clicking submit then this would be BIC.

In light of the above, `ref` change should instead be MAJOR, no?

Reply:
@melnikovi spoke with @tomreece and @jilu1, they both agree that changing `ref` would have to account for the most severe of test material changes, which would make it a `MAJOR`.

Reply:
Also regarding other attributes, `actionGroup` element only has `ref`, `stepKey` (identifier attribute) and `before/after` which are not relevant to our own versioning.